satellite/metabase/rangedloop/service.go

```diff
@@ -65,7 +77,76 @@ func (service *Service) Run(ctx context.Context) (err error) {
 func (service *Service) RunOnce(ctx context.Context) (err error) {
 	defer mon.Task()(&ctx)(&err)
 
-	// TODO
+	startTime := time.Now()
+	observerStates := []observerState{}
+
+	for _, obs := range service.observers {
+		err := obs.Start(ctx, startTime)
+		if err != nil {
+			return err
+		}
+
+		observerStates = append(observerStates, observerState{
+			observer: obs,
+			//rangeObservers: ranges,
+		})
+	}
+
+	rangeProviders, err := service.provider.CreateRanges(service.config.Parellelism, service.config.BatchSize)
+	if err != nil {
+		return err
+	}
+
+	group := errs2.Group{}
+	for rangeIndex, rangeProvider := range rangeProviders {
+		rangeObservers := []Partial{}
+		for i, observerState := range observerStates {
+			rangeObserver, err2 := observerState.observer.Fork(ctx)
+			if err2 != nil {
+				return err2
+			}
+			rangeObservers = append(rangeObservers, rangeObserver)
+			observerStates[i].rangeObservers = append(observerStates[i].rangeObservers, rangeObserver)
```

Comment:
can observerStates[i] be just observerState? then `for _, observerState := range observerStates {`

Reply:
That doesn't work, because Go iterates by value, not by reference. `observerState` is a copy, so the change with `append()` is not visible outside the loop.